### PR TITLE
provider/archive: Implementing archive provider

### PR DESCRIPTION
**Pull request by BSick7**
## Purpose
This introduces a provider to generate archive files from content, a file, or a directory of files.
The original impetus for this is the ability to package lambda functions on-the-fly without the need for specific zip utilities packaged in your repo.

This PR only includes support for zip files.

## Checklist
- [x] Implement archive provider
- [x] Implement archive_file resource
- [x] Test archive_file resource
- [x] Implement "zip" archiver
- [x] Test "zip" archiver
- [x] Add archive provider docs

### CODE REVIEW

**Comment by steve-jansen** [comment]
:+1: this will make it so much easier to work with `aws_lambda_finction` resources

**Comment by stack72** [comment]
So i've been testing this for a while now with lambda and it works well - i am going to merge it :) I believe it's a simple and effective way of getting going

Thanks @BSick7

**Comment by stack72** [comment]
```
% make testacc TEST=./builtin/providers/archive TESTARGS='-run=TestAccArchiveFile_'
==> Checking that code complies with gofmt requirements...
/Users/stacko/Code/go/bin/stringer
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/08/08 12:49:20 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/archive -v -run=TestAccArchiveFile_ -timeout 120m
=== RUN   TestAccArchiveFile_Basic
--- PASS: TestAccArchiveFile_Basic (0.04s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/archive	0.052s
```

File: website/source/docs/providers/archive/r/file.html.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+NOTE: One of `source_content_filename` (with `source_content`), `source_file`, or `source_dir` must be specified.
+
+* `type` - (required) The type of archive to generate.
+  NOTE: `archive` is supported.
```
Comment (glasser) [inline website/source/docs/providers/archive/r/file.html.md]: This should say `zip`, right?

Comment (BSick7) [inline website/source/docs/providers/archive/r/file.html.md]: Good find!  I guess I was a little aggressive when I renamed `zip` to `archive`.

Comment (BSick7) [inline website/source/docs/providers/archive/r/file.html.md]: https://github.com/hashicorp/terraform/pull/8373

**Comment by ask0n** [comment]
Maybe I have missed something, but how does it track, that source_file was changed? I don't find anything similar to source_code_hash of aws_lambda_function.

**Comment by ponyfleisch** [comment]
Hey all! Please take note of #9924, which implements a base_archive argument so you can keep all the dependencies of the lambda function in an archive and add individual files to that instead of cluttering the directory tree (and your repository)

Let me know if there is anything i can do to increase the chances of this ending up in terraform. I'm planning to release a module that makes use of this.

**Pull request merged**